File: Data Structures & Algorithms/binary-search/submission-0.java
class Solution {

    int bs(int[] nums, int i, int j, int target) {
        if(i == j) {
            if(nums[i] == target) {
                return i;
            } else {
                return -1;
            }
        } 
        int mid = i + ((j - i) / 2);

        if(nums[mid] == target) return mid;

        if(target <= nums[mid]) {
            return bs(nums, i, mid, target);
        } else {
            return bs(nums, mid+1, j, target);
        }
    }

    public int search(int[] nums, int target) {

        int size = nums.length;
        if(size == 1 && nums[0] == target){
            return 0;
        } 

        return bs(nums, 0, size-1, target);
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {

        let res = []

        candidates.sort((a,b) => a-b)

        const dfs = (i, curr, total) => {
            if(total == target) {
                res.push([...curr])
                return
            }
            
            if(i == candidates.length || total > target) return

            

            if(total + candidates[i] > target) return

            curr.push(candidates[i])
            dfs(i+1, curr, total + candidates[i]);
            curr.pop()

            while(candidates[i+1] == candidates[i] && i+1 < candidates.length) {
                i++;
            }
            dfs(i+1, curr, total);


        }

        dfs(0, [], 0);
        return res;
    }
}
